### PR TITLE
Fix race condition in DartVmServiceDebugProcessZ

### PR DESCRIPTION
**Pull request by jacob314**
Also fix race condition determining what Flutter extensions are available resulting in 
inspector toolbar buttons typically being disabled.

Also add code to be a bit paranoid about ensuring all updates to PerfService occur on the UI thread.
We could optimize that code a bit and only switch to the UI thread when about to update fields.

fixes https://github.com/flutter/flutter-intellij/issues/2071

### CODE REVIEW

File: third_party/intellij-plugins-dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceDebugProcessZ.java [hunk 1]
```diff
@@ -267,6 +269,39 @@ private void onConnectSucceeded(VmService vmService,
     final FlutterLaunchMode launchMode = FlutterLaunchMode.getMode(executionEnvironment);
     if (launchMode.supportsDebugConnection()) {
       myVmServiceWrapper.handleDebuggerConnected();
+
+      // TODO(jacobr): the following code is a workaround for
```
[the remark this answers is not in the document]
Comment (jacob314) [reply third_party/intellij-plugins-dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceDebugProcessZ.java]: done

**Comment by jacob314** [comment]
I wasn't satisfied by the fragile code and boilerplate in the initial pull request so I added an "EventStream" class borrowing ideas from the Dart stream class that makes the code a bit simpler.

File: src/io/flutter/utils/EventStream.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+/**
+ * Simple class for listening for a value that can be repeatedly updated.
+ * <p>
+ * The  benefit of this class instead of listening for events is you don't have
+ * to worry about missing values posted to the stream before your code started
+ * listening.
+ * The benefit of using this class over awaiting a Future is that the value can
+ * update multiple times.
+ * <p>
+ * New values to the EventStream can be posted on any thread to listen can be
```
Comment (devoncarew) [inline src/io/flutter/utils/EventStream.java]: some grammar fixes needed here -

File: src/io/flutter/utils/EventStream.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ * <p>
+ * The class is inspired by listen method on the Stream class in Dart.
+ */
+public class EventStream<T> {
+
+  protected final HashSet<StreamSubscription<T>> subscriptions = new LinkedHashSet<>();
+
+  private volatile T currentValue;
+
+  public EventStream(T initialValue) {
```
Comment (devoncarew) [inline src/io/flutter/utils/EventStream.java]: Perhaps also have a 

```
public EventStream() {
  currentValue = null;
}
```

constructor (and use this in the call sites that currently pass in null)?

File: third_party/intellij-plugins-dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceDebugProcessZ.java [hunk 4]
```diff
@@ -267,6 +269,39 @@ private void onConnectSucceeded(VmService vmService,
...
+                  });
+                }
+              }
+            });
+          }
+        }
+
+        @Override
+        public void onError(RPCError error) {
+          System.out.println("XXX got an error" + error);
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply third_party/intellij-plugins-dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceDebugProcessZ.java]: We should remove the System.out.println here -

File: src/io/flutter/utils/EventStream.java (hunk 3 above)
Comment (jacob314) [reply src/io/flutter/utils/EventStream.java]: done.

File: src/io/flutter/utils/EventStream.java (hunk 2 above)
Comment (jacob314) [reply src/io/flutter/utils/EventStream.java]: cleaned up a bit. feel free to send a follow up CL with to cleanup nits and polish the docs or api. this is the minimum for a useful Stream api. As a followup when time allows I'll adopt this API in more existing locations where it is simpler than what I am doing today.

**Pull request merged**